homeassistant/components/remote/xiaomi_miio.py

```diff
@@ -0,0 +1,255 @@
+
+    hidden = config.get(ATTR_HIDDEN)
+
+    xiaomi_miio_remote = XiaomiMiioRemote(
+        friendly_name, device, slot, timeout,
+        hidden, config.get(CONF_COMMANDS))
+
+    hass.data[PLATFORM][host] = xiaomi_miio_remote
+
+    async_add_devices([xiaomi_miio_remote])
+
+    @asyncio.coroutine
+    def async_service_handler(service):
+        """Handle a learn command."""
+        if service.service != SERVICE_LEARN:
+            _LOGGER.error("We should not handle service: %s", service.service)
+            return
+
+        entity_id = service.data.get(ATTR_ENTITY_ID)
+        entity = None
+        for remote in hass.data[PLATFORM].values():
+            if remote.entity_id == entity_id:
+                entity = remote
+
+        if not entity:
+            _LOGGER.error("entity_id: '%s' not found", entity_id)
+            return
+
+        device = entity.device
+
+        slot = service.data.get(CONF_SLOT, entity.slot)
+
+        yield from hass.async_add_job(device.learn, slot)
+
+        timeout = service.data.get(CONF_TIMEOUT, entity.timeout)
+
+        _LOGGER.info("Press the key you want Home Assistant to learn")
+        start_time = utcnow()
+        while (utcnow() - start_time) < timedelta(seconds=timeout):
+            message = yield from hass.async_add_job(
+                device.read, slot)
+            _LOGGER.debug("Message recieved from device: '%s'", message)
+
+            if 'code' in message and message['code']:
+                log_msg = "Received command is: {}".format(message['code'])
+                _LOGGER.info(log_msg)
+                hass.components.persistent_notification.async_create(
+                    log_msg, title='Xiaomi Miio Remote')
+                return
+
+            if ('error' in message and
+                    message['error']['message'] == "learn timeout"):
+                yield from hass.async_add_job(device.learn, slot)
+
+            yield from asyncio.sleep(1, loop=hass.loop)
+
+        _LOGGER.error("Timeout. No infrared command captured")
+        hass.components.persistent_notification.async_create(
+            "Timeout. No infrared command captured",
+            title='Xiaomi Miio Remote')
+
+    hass.services.async_register(DOMAIN, SERVICE_LEARN, async_service_handler,
+                                 schema=LEARN_COMMAND_SCHEMA)
+
+
+class XiaomiMiioRemote(RemoteDevice):
+    """Representation of a Xiaomi Miio Remote device."""
+
+    def __init__(self, friendly_name, device,
+                 slot, timeout, hidden, commands):
+        """Initialize the remote."""
+        self._name = friendly_name
+        self._device = device
+        self._is_hidden = hidden
+        self._slot = slot
+        self._timeout = timeout
+        self._state = False
+        self._commands = commands
+
+    @property
+    def name(self):
+        """Return the name of the remote."""
+        return self._name
+
+    @property
+    def device(self):
```

Comment:
If the property isn't needed just make a regular instance attribute of it instead.